### PR TITLE
Resolving generics and nested functions

### PR DESCRIPTION
**Pull request by lukasholzer**
This pull request is still work in progress!

This is a rewrite the whole code-analyzer in case that it was necessary to reimplement the parsed AST structure and the visitor. Now every Typescript node gets parsed and resolved.

### CODE REVIEW

**Comment by lukasholzer** [comment]
Following Code is now able to parse!

```typescript
export type Constructor<T> = new(...args: any[]) => T;

interface I1 {
  i: boolean;
}

interface I2  {
  i2: number;
}

function f<T extends Constructor<I1>>(a: T): T & Constructor<I2> {
  return a as any;
}

class X { 
  i: boolean;
  x: boolean;
}


const XI2 = f(X);
```

**Comment by lukasholzer** [comment]
```typescript
type Constructor<T> = new(...args: any[]) => T;

interface CanDisable {
  disabled: boolean;
}

interface CanColor<P extends Partial<DtThemePalette>> {
  color: P;
}

interface HasElementRef {
  _elementRef: ElementRef;
}

type DtThemePalette = 'main' | 'accent' | 'warning' | 'error' | 'cta' | 'recovered' | 'neutral' | undefined;

function mixinColor<T extends Constructor<HasElementRef>, P extends Partial<DtThemePalette>>(
  base: T, defaultColor?: P): Constructor<CanColor<P>> & T {
  return class extends base {
    private _color: P;

    get color(): P { return this._color; }
    set color(value: P) {
      const colorPalette = value || defaultColor as P;

      if (colorPalette !== this._color) {
        this._color = colorPalette;
      }
    }

    constructor(...args: any[]) {
      super(...args);
      this.color = defaultColor as P;
    }
  };
}

function mixinDisabled<T extends Constructor<{}>>(base: T): Constructor<CanDisable> & T {
  return class extends base {
    private _disabled = false;

    get disabled(): boolean { return this._disabled; }
    set disabled(value: boolean) { this._disabled = value; }

    constructor(...args: any[]) { super(...args); }
  };
}


type DtButtonThemePalette = 'main' | 'warning' | 'cta';

class DtButtonBase {
  constructor(public _elementRef: ElementRef) { }
}
export type ButtonVariant = 'primary' | 'secondary' | 'nested';
const defaultVariant = 'primary';


const _DtButtonMixinBase = mixinDisabled(mixinColor<Constructor<DtButtonBase>, DtButtonThemePalette>(DtButtonBase, 'main'));

@Component({
  moduleId: module.id,
  selector: `button[dt-button], button[dt-icon-button]`,
  exportAs: 'dtButton',
  host: {
    'class': 'dt-button',
    '[class.dt-icon-button]': '_isIconButton',
    '[disabled]': 'disabled || null',
  },
  templateUrl: 'button.html',
  styleUrls: ['button.scss'],
  inputs: ['disabled', 'color'],
  encapsulation: ViewEncapsulation.None,
  preserveWhitespaces: false,
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class DtButton extends _DtButtonMixinBase
  implements OnDestroy, AfterContentInit, CanDisable, CanColor<DtButtonThemePalette>, HasElementRef {

  @Input()
  get variant(): ButtonVariant { return this._variant; }
  set variant(value: ButtonVariant) {
    const variant = value || defaultVariant;
    if (variant !== this._variant) {
      this._replaceCssClass(variant, this._variant);
      this._variant = variant;
    }
  }
  private _variant: ButtonVariant;
  private _iconChangesSub: Subscription = NEVER.subscribe();

  @ContentChildren(DtIcon) _icons: QueryList<DtIcon>;

  /** @internal Whether the button is icon only. */
  _isIconButton = false;

  constructor(
    elementRef: ElementRef,
    private _focusMonitor: FocusMonitor,
    private _renderer: Renderer2,
    private _changeDetectorRef: ChangeDetectorRef
  ) {
    super(elementRef);

    // Set the default variant to trigger the setters.
    this.variant = defaultVariant;

    this._isIconButton = this._getHostElement().hasAttribute('dt-icon-button');

    this._focusMonitor.monitor(this._elementRef.nativeElement, true);
  }

  ngAfterContentInit(): void {
    // We need to set markForCheck manually on every icons change
    // so that the template can determine if the icon container
    // should be shown or not
    this._iconChangesSub = this._icons.changes
      .subscribe(() => { this._changeDetectorRef.markForCheck(); });
  }

  ngOnDestroy(): void {
    this._focusMonitor.stopMonitoring(this._elementRef.nativeElement);
    this._iconChangesSub.unsubscribe();
  }

  /** Focuses the button. */
  focus(): void {
    this._getHostElement().focus();
  }

  /** Retrieves the native element of the host. */
  private _getHostElement(): HTMLElement {
    return this._elementRef.nativeElement;
  }

  private _replaceCssClass(newClass?: string, oldClass?: string): void {
    replaceCssClass(this._elementRef, `dt-button-${oldClass}`, `dt-button-${newClass}`, this._renderer);
  }

}

````


results in following output:

```json
[ [ { name: 'DtButton',
      angularComponent: true,
      decorator:
       { selector: '`button[dt-button], button[dt-icon-button]`',
         exportAs: '\'dtButton\'',
         inputs: [ '\'disabled\'', '\'color\'' ] },
      members:
       [ { type: 'property', key: 'disabled', value: 'true' },
         { type: 'property',
           key: 'color',
           value: [ '\'main\'', '\'warning\'', '\'cta\'' ] },
         { type: 'property',
           key: 'variant',
           value: [ '\'primary\'', '\'secondary\'', '\'nested\'' ] } ] } ] ]
```

File: packages/code-analyzer/src/utils/get-node-tags.ts [hunk 1]
```diff
@@ -1,27 +1,57 @@
...
+        case ts.SyntaxKind.PrivateKeyword:
+          tags.push('private'); break;
+        case ts.SyntaxKind.PublicKeyword:
+          tags.push('public'); break;
+        case ts.SyntaxKind.ProtectedKeyword:
+          tags.push('protected'); break;
+        case ts.SyntaxKind.StaticKeyword:
+          tags.push('static'); break;
+        case ts.SyntaxKind.ReadonlyKeyword:
+          tags.push('readonly'); break;
```
Comment (tomheller) [inline packages/code-analyzer/src/utils/get-node-tags.ts]: no default case. 😢

File: packages/code-analyzer/src/utils/get-symbol-name.ts [hunk 2]
```diff
@@ -8,38 +8,56 @@ const log = new Logger();
...
+    return node.expression.text;
+  }
+
+  switch (node.kind) {
+    case ts.SyntaxKind.VariableStatement:
+      return getSymbolName(node.declarationList);
+    case ts.SyntaxKind.VariableDeclarationList:
+      return node.declarations.map(declaration => getSymbolName(declaration)).join(', ');
+    // call expression that does not match the if case before.
+    case ts.SyntaxKind.CallExpression:
```
Comment (tomheller) [inline packages/code-analyzer/src/utils/get-symbol-name.ts]: wouldn't that be the default case... 51-56

**Pull request merged**